### PR TITLE
Extend CubicSupercell transformation to also be able to look for orthorhombic cells

### PR DESCRIPTION
**Pull request by JaGeo**
## Summary

Added code to also allow to search for orthorhombic supercells. Here, I also introduce the option to check for a maximum_length of the cells

@QuantumChemist will now take over and write the tests for this class and fix the linting etc. :)

### CODE REVIEW

File: src/pymatgen/transformations/advanced_transformations.py [hunk 1]
```diff
@@ -1488,64 +1499,92 @@ def apply_transformation(self, structure: Structure) -> Structure:
...
+                    or np.all(np.absolute(np.array(superstructure.lattice.angles) - 90) < self.angle_tolerance)
+                ):
+                    return superstructure
+
+                self.transformation_matrix=transformation_matrix
+                # Increase threshold until proposed supercell meets requirements
+                target_sc_size += self.step_size
+                self.check_exceptions(length_vecs, n_atoms)
+            raise AttributeError("Unable to find cubic supercell")
+        # boolean for if a sufficiently large supercell has been created
```
Comment (JaGeo) [inline src/pymatgen/transformations/advanced_transformations.py:1528]:
```suggestion

```

File: src/pymatgen/transformations/advanced_transformations.py [hunk 2]
```diff
@@ -1488,64 +1498,112 @@ def apply_transformation(self, structure: Structure) -> Structure:
...
+                lat_vecs, structure, target_sc_lat_vecs
+            )
+            print(superstructure.lattice)
             # Check if constraints are satisfied
+            print(n_atoms)
             if (
                 np.min(np.linalg.norm(length_vecs, axis=1)) >= self.min_length
                 and self.min_atoms <= n_atoms <= self.max_atoms
             ) and (
                 not self.force_90_degrees
```
Comment (QuantumChemist) [inline src/pymatgen/transformations/advanced_transformations.py:1548]: @JaGeo , if I set `force_90_degrees` to True, I never leave the while loop

Comment (JaGeo) [reply src/pymatgen/transformations/advanced_transformations.py:1548]: That's not true. There is an additional statement here.

Comment (JaGeo) [reply src/pymatgen/transformations/advanced_transformations.py:1548]: ` or np.all(np.absolute(np.array(superstructure.lattice.angles) - 90) < self.angle_tolerance)`

Comment (QuantumChemist) [reply src/pymatgen/transformations/advanced_transformations.py:1546]: the unit test runs forever without stopping when I set force_90_degrees to True for the orthorhombic supercell generation. the test finalized for the cubic supercell transformation.

Comment (JaGeo) [reply src/pymatgen/transformations/advanced_transformations.py:1546]: but not because of this. Can you test a bit more with min/max lengths?

Comment (QuantumChemist) [reply src/pymatgen/transformations/advanced_transformations.py:1546]: Oh sorry, my bad, I see now that this is even outside the while loop.

Comment (JaGeo) [reply src/pymatgen/transformations/advanced_transformations.py:1546]: No, this is really the stopping criterion but it should still stop or run into the attribute error if there is no cell like this.

Comment (QuantumChemist) [reply src/pymatgen/transformations/advanced_transformations.py:1546]: we can resolve this conversation, as step_size was the crucial part here

Comment (JaGeo) [reply src/pymatgen/transformations/advanced_transformations.py:1546]: Maybe add a hint to the doc that stepsize needs to be larger for the orthorhombic cells

Comment (QuantumChemist) [reply src/pymatgen/transformations/advanced_transformations.py]: ok, done :)

**Comment by QuantumChemist** [comment]
The failing linting does not stem from our changes.

**Comment by JaGeo** [comment]
Once the other tests pass, this is ready for review

**Comment by JaGeo** [comment]
@shyuep @mkhorton @janosh this pull request is ready for review. The linting issues show up in different parts of the code.

**Comment by DanielYang59** [comment]
There are many mysteries in the world of linters, including why they sometimes flag untouched parts of the code out of the blue 🤣 .

**src/pymatgen/analysis/local_env.py:3051:32:**
```diff
- if (m != j) and (m != k) and (not flag_xaxis):
+ if m not in {j, k} and (not flag_xaxis):
```

**src/pymatgen/io/aims/parsers.py:333:12:**
```diff
- if (line_start == LINE_NOT_FOUND) or (line_end == LINE_NOT_FOUND) or (line_end - line_start != n_kpts):
+ if LINE_NOT_FOUND in {line_start, line_end} or (line_end - line_start != n_kpts):
```

src/pymatgen/io/vasp/outputs.py:1707:25:
```diff
- if (
-                        tag == "eigenvalues_kpoints_opt"
-                        or tag == "projected_kpoints_opt"
-                        or (tag == "dos" and elem.attrib.get("comment") == "kpoints_opt")
-                    ):
+ if (
+                        tag in { "eigenvalues_kpoints_opt", "projected_kpoints_opt"}
+                        or (tag == "dos" and elem.attrib.get("comment") == "kpoints_opt")
+                    ):
```

**Comment by JaGeo** [comment]
@DanielYang59 can you try again? You should now be allowed to do this

**Comment by DanielYang59** [comment]
Beautiful! Thanks for granting the access. Actually I was not looking for writing access to your branch, but to open a PR where your fork repo is the base repo, but I cannot find your fork as the target repo for some reason.

**Comment by JaGeo** [comment]
@DanielYang59 i am very confused. You mean this branch https://github.com/JaGeo/pymatgen/tree/orthorhombic_supercell ?

**Comment by DanielYang59** [comment]
Sorry for the confusion. For some reason I can see your fork now (wasn't showing up before):
image

Everything is working fine now! Thanks!

**Comment by JaGeo** [comment]
@DanielYang59 thank you 😀.

I hope someone will then merge our backward compatible code soon!

**Comment by DanielYang59** [comment]
No problem!

**Comment by QuantumChemist** [comment]
> Beautiful! Thanks for granting the access. Actually I was not looking for writing access to your fork, but to open a PR where your fork is the base repo, but I cannot find yours as the target repo for some reason.

I had the same problem yesterday btw. I worked around it by starting the PR in the JaGeo repo. But good everything worked in the end.

**Comment by shyuep** [comment]
Merged. Sometimes lint errors come up on other parts due to update of the linting software versions.

**Comment by JaGeo** [comment]
Thank you @shyuep

**Pull request merged**